pkg/controller/deployment_rolling.go

```diff
@@ -236,3 +251,90 @@ func (dc *controller) scaleDownOldMachineSetsForRollingUpdate(allISs []*v1alpha1
 
 	return totalScaledDown, nil
 }
+
+// taintNodesBackingMachineSets taints all nodes backing the machineSets
+func (dc *controller) taintNodesBackingMachineSets(MachineSets []*v1alpha1.MachineSet) error {
+
+	for _, machineSet := range MachineSets {
+
+		if machineSet.Annotations["PreferNoSchedule"] != "" {
+			continue
+		}
+
+		glog.V(3).Infof("Trying to taint MachineSet object %q with prefer-no-schedule to avoid scheduling of pods", machineSet.Name)
+		selector, err := metav1.LabelSelectorAsSelector(machineSet.Spec.Selector)
+		if err != nil {
+			return err
+		}
+
+		// list all machines to include the machines that don't match the ms`s selector
+		// anymore but has the stale controller ref.
+		// TODO: Do the List and Filter in a single pass, or use an index.
+		filteredMachines, err := dc.machineLister.List(labels.Everything())
+		if err != nil {
+			return err
+		}
+		// NOTE: filteredMachines are pointing to objects from cache - if you need to
+		// modify them, you need to copy it first.
+		filteredMachines, err = dc.claimMachines(machineSet, selector, filteredMachines)
+		if err != nil {
+			return err
+		}
+
+		taints := v1.Taint{
+			Key:    "old-machine-set",
+			Value:  "True",
+			Effect: "PreferNoSchedule",
+		}
+
+		// Iterate through all machines and place the PreferNoSchedule taint
+		// to avoid scheduling on older machines
+		for _, machine := range filteredMachines {
+			if machine.Labels["node"] != "" {
+				err = AddOrUpdateTaintOnNode(
+					dc.targetCoreClient,
+					machine.Labels["node"],
+					&taints,
+				)
+				if err != nil {
+					glog.Errorf("Node tainting failed for node: %s, %s", machine.Labels["node"], err)
+					return err
```

Comment:
In case the error occurs after labelling some (not all) machines in a machine set, we return error. Does it fail the whole rolling update operation without retry? If yes, then we might also need to remove the successful taints/annotations before returning. WDYT?
Same goes for all other places where we return because of failure - not just in this function, but in subsequent function calls too.

Reply:
This is something important that you have brought up here and I think it makes sense. 

- Does it fail the whole rolling update operation without retry? 
Yes it does fail the while rolling update for then. It retries it later in the next machineDeployment reconciliation which is forced every 10 mins.

-  If yes, then we might also need to remove the successful taints/annotations before returning
I was thinking of an alternate approach where even if the tainting operation fails, we continue the machine rolling update (as it is meant to carry forward even if this step fails). The tainting would then be re-tried in the next machineDeployment reconciliation (which should be in the next few seconds as the machine deployment would be updated soon)?

Reply:
yes. that makes sense too.

Reply:
I shall do that then.